Add mutation events tests for ContainerNode::replaceChildren and refactor the code

rniwa opened the pull request:
#### c177199a72c0d38878dcffa142e64ffe263e822e
<pre>
Add mutation events tests for ContainerNode::replaceChildren and refactor the code
<a href="https://bugs.webkit.org/show_bug.cgi?id=261016">https://bugs.webkit.org/show_bug.cgi?id=261016</a>

Reviewed by Chris Dumez.

Added some tests for mutating DOM during replaceChildren&apos;s tree mutating operations.

Also refactor the code in ContainerNode::replaceChildren by merging some aspects of
Node::convertNodesOrStringsIntoNode into the function so that all node removal &amp;
insertion the logic will be in one place.

* LayoutTests/fast/dom/replace-children-mutation-events-expected.txt: Added.
* LayoutTests/fast/dom/replace-children-mutation-events.html: Added.
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::replaceChildren):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::convertNodesOrStringsIntoNodeVector): Extracted from convertNodesOrStringsIntoNode.
(WebCore::Node::convertNodesOrStringsIntoNode):
* Source/WebCore/dom/Node.h:

Canonical link: <a href="https://commits.webkit.org/267565@main">https://commits.webkit.org/267565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d3520b9c701a5a475a1938c63eaf83ad61faf24

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17001 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17324 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18784 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15908 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17191 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17463 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17202 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17547 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14735 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19599 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14789 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22142 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15774 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15596 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19921 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16190 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13727 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15350 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15269 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19714 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2091 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16023 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/fast/dom/replace-children-mutation-events.html:
```diff
@@ -0,0 +1,26 @@
+<body>
+<script src="../../resources/js-test.js"></script>
+<script>
+
```
This test is missing a `description()`.

rniwa replied:
Fixed.

cdumez commented on Source/WebCore/dom/Node.h:
```diff
@@ -710,6 +710,7 @@ class Node : public EventTarget {
     void updateAncestorsForStyleRecalc();
     void markAncestorsForInvalidatedStyle();
 
+    Vector<Ref<Node>, 1> convertNodesOrStringsIntoNodeList(FixedVector<NodeOrString>&&);
```
I would suggest using NodeVector instead of NodeList. NodeList is a very specific DOM type and I thought this was what was being returned at first.

rniwa replied:
Okay, I did debate between naming that vs. this so I'd change it.

cdumez commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -587,11 +587,19 @@ ExceptionOr<RefPtr<Node>> Node::convertNodesOrStringsIntoNode(FixedVector<NodeOr
+
+ExceptionOr<RefPtr<Node>> Node::convertNodesOrStringsIntoNode(FixedVector<NodeOrString>&& nodeOrStringVector)
+{
+    auto nodeList = convertNodesOrStringsIntoNodeList(WTFMove(nodeOrStringVector));
```
nodeVector

rniwa replied:
Will fix.

cdumez reviewed (approved):
LGTM though we should closely monitor perf bots after landing.

rniwa commented on LayoutTests/fast/dom/replace-children-mutation-events.html:
```diff
@@ -0,0 +1,26 @@
+shouldThrowErrorName('parent.replaceChildren(span)', 'NotFoundError');
+
+oldParent.addEventListener('DOMNodeInserted', () => document.body.appendChild(i), {once: true});
+shouldThrowErrorName('span.replaceChildren(oldParent, i, "other")', 'HierarchyRequestError');
```
Oh oops, I have a bug here. I'm not testing mutation events because i is span's parent at this point.  Will fix the test as well.

rniwa committed:
Add mutation events tests for ContainerNode::replaceChildren and refactor the code
https://bugs.webkit.org/show_bug.cgi?id=261016

Reviewed by Chris Dumez.

Added some tests for mutating DOM during replaceChildren's tree mutating operations.

Also refactor the code in ContainerNode::replaceChildren by merging some aspects of
Node::convertNodesOrStringsIntoNode into the function so that all node removal &
insertion the logic will be in one place.

* LayoutTests/fast/dom/replace-children-mutation-events-expected.txt: Added.
* LayoutTests/fast/dom/replace-children-mutation-events.html: Added.
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::replaceChildren):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::convertNodesOrStringsIntoNodeVector): Extracted from convertNodesOrStringsIntoNode.
(WebCore::Node::convertNodesOrStringsIntoNode):
* Source/WebCore/dom/Node.h:

Canonical link: https://commits.webkit.org/267565@main
